### PR TITLE
ipywidgets UI create delete

### PR DESCRIPTION
**Pull request by Bobbins228**
# Issue link

Closes: [RHOAIENG-12523](https://issues.redhat.com/browse/RHOAIENG-12523)
# What changes have been made

* Added ipywidgets to act as buttons for the `cluster.up()` and `cluster.down()` methods
* Added print statements to confirm cluster creation/deletion
* Added unit tests for buttons
* Added wait_ready Check Box

image
Happy days :) 
# Verification steps

### Notebook server ODH/RHOAI/Local
* Clone this repository with `git clone https://github.com/project-codeflare/codeflare-sdk.git`
* Checkout this PR's branch
* Run `poetry build` - install if needed (`pip install poetry`)
* Run `pip install --force-reinstall dist/codeflare_sdk-0.0.0.dev0-py3-none-any.whl` 
* Restart your notebook kernel
## Testing
* Run through any example notebook
* On the `ClusterConfiguration` step the ui buttons for cluster up/down should appear
* Verify the buttons work as expected
* Run the `ClusterConfiguration` as a script and the buttons should not appear.
* Run in a RHOAI notebook and in a VSCode local notebook and verify buttons are created

## Checks
- [ ] I've made sure the tests are passing. 
- Testing Strategy
   - [ ] Unit tests
   - [ ] Manual tests
   - [ ] Testing is not required for this change

### CODE REVIEW

File: src/codeflare_sdk/cluster/cluster.py [hunk 1]
```diff
@@ -71,6 +74,11 @@ def __init__(self, config: ClusterConfiguration):
         self.app_wrapper_yaml = self.create_app_wrapper()
         self._job_submission_client = None
         self.app_wrapper_name = self.config.name
+        if (
+            "PYDEVD_IPYTHON_COMPATIBLE_DEBUGGING" in os.environ
+            or "JPY_SESSION_NAME" in os.environ
+        ):  # If running Jupyter NBs in VsCode or RHOAI/ODH display UI buttons
+            self._cluster_up_down_buttons()
```
Comment (KPostOffice) [inline src/codeflare_sdk/cluster/cluster.py:81]: Can you pull the if statement into a utility function so it becomes like:
```suggestion
        if running_in_jlab():
            self._cluster_up_down_buttons()
```
We'll end up needing this in a lot of places for the new functionality

File: tests/unit_test.py [hunk 2]
```diff
@@ -2873,6 +2876,50 @@ def test_cluster_config_deprecation_conversion(mocker):
...
+        "ipywidgets.HBox"
+    ), patch.object(cluster, "up") as mock_up, patch.object(
+        cluster, "down"
+    ) as mock_down:
+        # Create mock button instances
+        mock_up_button = MagicMock()
+        mock_down_button = MagicMock()
+
+        # Ensure the mock Button class returns the mock button instances in sequence
+        MockButton.side_effect = [mock_down_button, mock_up_button]
```
Comment (KPostOffice) [inline tests/unit_test.py:2907]: This is a tiny nit, but shouldn't the list order be swapped?

**Review by KPostOffice — COMMENTED** [review]
Looks great, I've left a couple of comments

File: src/codeflare_sdk/cluster/cluster.py (hunk 1 above)
Comment (Bobbins228) [reply src/codeflare_sdk/cluster/cluster.py:81]: Good idea will update :+1:

File: src/codeflare_sdk/cluster/cluster.py [hunk 3]
```diff
@@ -156,11 +161,42 @@ def up(self):
...
             return _kube_api_error_handling(e)
 
+    def _cluster_up_down_buttons(self):
+        delete_button = widgets.Button(
+            description="Cluster Down",
+            icon="trash",
+        )
+        up_button = widgets.Button(
+            description="Cluster Up",
+            icon="play",
```
Comment (ChristianZaccaria) [inline src/codeflare_sdk/cluster/cluster.py:180]: You could add a `tooltip` and include a small description on what the button does. The user will be able to see the message when hovering over the button.
Reference: https://ipywidgets.readthedocs.io/en/latest/examples/Widget%20List.html#togglebutton

**Review by ChristianZaccaria — DISMISSED** [review]
Great work! Just one small nitpick

**Review by ChristianZaccaria — COMMENTED** [review]
Should we think of moving UI related items into their own file? - And keep strictly functionality in the `cluster.py` file?

File: src/codeflare_sdk/cluster/cluster.py (hunk 3 above)
Comment (Bobbins228) [reply src/codeflare_sdk/cluster/cluster.py:180]: That's pretty cool. I'll add descriptions

**Comment by Bobbins228** [comment]
@ChristianZaccaria I had a similar thought when reading through the design doc. 
I think we should go for it as to not clog the cluster.py file with the widgets.

**Comment by Bobbins228** [comment]
@ChristianZaccaria On second thought that may not be the best for this particular widget. 
It requires the Cluster Object for the `up()` and `down()` methods so it is seemingly the only widget that should live in `cluster.py`.

**Comment by ChristianZaccaria** [comment]
@Bobbins228 perhaps the UI itself (only the buttons for example) can be separated into it's own file, and the functionality (`up()` and `down()`) remain in the cluster.py file.

File: src/codeflare_sdk/cluster/widgets.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+# limitations under the License.
+
+"""
+The widgets sub-module contains the ui widgets created using the ipywidgets package.
+"""
+import ipywidgets as widgets
+import os
+
+
+def cluster_up_down_buttons(appwrapper: bool) -> widgets.Button:
```
Comment (KPostOffice) [inline src/codeflare_sdk/cluster/widgets.py:22]: Can you add the cluster as an arg here and add the onclick functionality

**Review by KPostOffice — COMMENTED** [review]
One change, other than that I think it's g2g

File: src/codeflare_sdk/cluster/widgets.py (hunk 4 above)
Comment (Bobbins228) [reply src/codeflare_sdk/cluster/widgets.py:22]: Great point Kevin we can remove all Ipywigets mentions from `cluster.py` this way. Updating the PR.

File: src/codeflare_sdk/cluster/widgets.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
+# Copyright 2022 IBM, Red Hat
```
Comment (KPostOffice) [inline src/codeflare_sdk/cluster/widgets.py:1]: Last change I promise
```suggestion
# Copyright 2024 IBM, Red Hat
```

**Review by KPostOffice — APPROVED** [review]
/lgtm

**Review by Fiona-Waters — COMMENTED** [review]
/lgtm

**Pull request merged**